app/Http/Controllers/Api/VolumeController.php

```diff
@@ -144,4 +157,305 @@ public function update(UpdateVolume $request)
+                            $original['annotation_id'] = $newAnnotationId;
+                            unset($original['id']);
+                            $insertData[] = $original;
+                        }
+                    }
+                }
+                ImageAnnotationLabel::insert($insertData);
+            });
+    }
+
+    /**
+     * Copies (selected) image labels from given volume to volume copy.
+     *
+     * @param Volume $volume
+     * @param Volume $copy
+     * @param int[] $selectedImageIds
+     **/
+    private function copyImageLabels($volume, $copy, $selectedImageIds)
+    {
+        $oldImages = empty($selectedImageIds) ? $volume->images()->orderBy('id')->with('labels')->get() :
+            Image::whereIn('id', $selectedImageIds)->orderBy('id')->with('labels')->get();
+        $newImageIds = $copy->images()->orderBy('id')->get()->pluck('id');
+
+        foreach ($oldImages as $imageIdx => $oldImage) {
+            $newImageId = $newImageIds[$imageIdx];
+            $oldImage->labels->map(function ($oldLabel) use ($newImageId) {
+                $origin = $oldLabel->getRawOriginal();
+                $origin['image_id'] = $newImageId;
+                unset($origin['id']);
+                return $origin;
+            })->chunk(10000)->each(function ($chunk) {
+                ImageLabel::insert($chunk->toArray());
+            });
+        }
+    }
+
+    /**
+     * Copies (selected) videos from given volume to volume copy.
+     *
+     * @param Volume $volume
+     * @param Volume $copy
+     * @param int[] $selectedVideoIds
+     **/
+    private function copyVideos($volume, $copy, $selectedVideoIds)
+    {
+        // copy video references
+        $videos = $volume->videos()
+            ->when(!empty($videos), function ($query) use ($selectedVideoIds) {
+                $query->whereIn('id', $selectedVideoIds);
+            })->get();
+
+        $videos->map(function ($video) use ($copy) {
+            $original = $video->getRawOriginal();
+            $original['volume_id'] = $copy->id;
+            $original['uuid'] = (string)Uuid::uuid4();
+            unset($original['id']);
+            return $original;
+        })->chunk(10000)->each(function ($chunk) {
+            Video::insert($chunk->toArray());
+        });
+
+    }
+
+    /**
+     * Copies (selected) video annotations and annotation labels from given volume to volume copy.
+     *
+     * @param Volume $volume
+     * @param Volume $copy
+     * @param int[] $selectedVideoIds
+     **/
+    private function copyVideoAnnotation($volume, $copy, $selectedVideoIds)
+    {
+        //TODO: use selected videoIds
+        $chunkSize = 100;
+        $newVideoIds = $copy->videos()->orderBy('id')->pluck('id');
+        $volume->videos()
+            ->orderBy('id')
+            ->with('annotations.labels')
+            // Note the use of chunkById() instead of previously eachById().
```

Comment:
```suggestion
```suggestion
            // This is an optimized implementation to clone the annotations with only few database
            // queries. There are simpler ways to implement this but they can be ridiculously 
            // inefficient.
```
```